### PR TITLE
`azurerm_iothub` -  split create and update function

### PR DESCRIPTION
**Pull request by ziyeqf**
The combained `createupdate` function also cause a bug when there is a endpoint attached to the iothub which is not supported by `azurerm_iothub`. So splitting the function also helps fixing the bug.

Then a thought about `endpoint` block, the scheme of it is a comabination of `azurerm_iothub_endpoint_xxx`, it's a bit complex, and duplicate with those resources, so maybe we can deprecate that block and let users only use `azurerm_iothub_endpoint_xxx` resources?

Test
---
```
❯ tftest iothub TestAccIotHub_                                  
=== RUN   TestAccIotHub_basic
=== PAUSE TestAccIotHub_basic
=== RUN   TestAccIotHub_networkRulesSet
=== PAUSE TestAccIotHub_networkRulesSet
=== RUN   TestAccIotHub_requiresImport
=== PAUSE TestAccIotHub_requiresImport
=== RUN   TestAccIotHub_standard
=== PAUSE TestAccIotHub_standard
=== RUN   TestAccIotHub_customRoutes
=== PAUSE TestAccIotHub_customRoutes
=== RUN   TestAccIotHub_enrichments
=== PAUSE TestAccIotHub_enrichments
=== RUN   TestAccIotHub_removeEndpointsAndRoutes
=== PAUSE TestAccIotHub_removeEndpointsAndRoutes
=== RUN   TestAccIotHub_fileUpload
=== PAUSE TestAccIotHub_fileUpload
=== RUN   TestAccIotHub_fileUploadAuthenticationTypeUserAssignedIdentity
=== PAUSE TestAccIotHub_fileUploadAuthenticationTypeUserAssignedIdentity
=== RUN   TestAccIotHub_fileUploadAuthenticationTypeUpdate
=== PAUSE TestAccIotHub_fileUploadAuthenticationTypeUpdate
=== RUN   TestAccIotHub_withDifferentEndpointResourceGroup
=== PAUSE TestAccIotHub_withDifferentEndpointResourceGroup
=== RUN   TestAccIotHub_fallbackRoute
=== PAUSE TestAccIotHub_fallbackRoute
=== RUN   TestAccIotHub_publicAccess
=== PAUSE TestAccIotHub_publicAccess
=== RUN   TestAccIotHub_minTLSVersion
=== PAUSE TestAccIotHub_minTLSVersion
=== RUN   TestAccIotHub_LocalAuth
=== PAUSE TestAccIotHub_LocalAuth
=== RUN   TestAccIotHub_cloudToDevice
=== PAUSE TestAccIotHub_cloudToDevice
=== RUN   TestAccIotHub_identitySystemAssigned
=== PAUSE TestAccIotHub_identitySystemAssigned
=== RUN   TestAccIotHub_identitySystemAssignedUserAssigned
=== PAUSE TestAccIotHub_identitySystemAssignedUserAssigned
=== RUN   TestAccIotHub_identityUserAssigned
=== PAUSE TestAccIotHub_identityUserAssigned
=== RUN   TestAccIotHub_identityUpdate
=== PAUSE TestAccIotHub_identityUpdate
=== RUN   TestAccIotHub_endpointAuthenticationTypeUserAssignedIdentity
=== PAUSE TestAccIotHub_endpointAuthenticationTypeUserAssignedIdentity
=== RUN   TestAccIotHub_endpointAuthenticationTypeUpdate
=== PAUSE TestAccIotHub_endpointAuthenticationTypeUpdate
=== CONT  TestAccIotHub_basic
=== CONT  TestAccIotHub_fallbackRoute
=== CONT  TestAccIotHub_removeEndpointsAndRoutes
=== CONT  TestAccIotHub_fileUploadAuthenticationTypeUpdate
=== CONT  TestAccIotHub_withDifferentEndpointResourceGroup
=== CONT  TestAccIotHub_endpointAuthenticationTypeUpdate
=== CONT  TestAccIotHub_fileUploadAuthenticationTypeUserAssignedIdentity
=== CONT  TestAccIotHub_fileUpload
--- PASS: TestAccIotHub_fallbackRoute (345.60s)
=== CONT  TestAccIotHub_enrichments
--- PASS: TestAccIotHub_basic (347.77s)
=== CONT  TestAccIotHub_requiresImport
--- PASS: TestAccIotHub_fileUploadAuthenticationTypeUserAssignedIdentity (472.11s)
=== CONT  TestAccIotHub_identitySystemAssigned
--- PASS: TestAccIotHub_fileUpload (479.68s)
=== CONT  TestAccIotHub_endpointAuthenticationTypeUserAssignedIdentity
--- PASS: TestAccIotHub_withDifferentEndpointResourceGroup (516.12s)
=== CONT  TestAccIotHub_identityUpdate
--- PASS: TestAccIotHub_requiresImport (337.82s)
=== CONT  TestAccIotHub_identityUserAssigned
--- PASS: TestAccIotHub_identitySystemAssigned (333.03s)
=== CONT  TestAccIotHub_identitySystemAssignedUserAssigned
--- PASS: TestAccIotHub_fileUploadAuthenticationTypeUpdate (822.61s)
=== CONT  TestAccIotHub_LocalAuth
--- PASS: TestAccIotHub_enrichments (535.28s)
=== CONT  TestAccIotHub_cloudToDevice
--- PASS: TestAccIotHub_removeEndpointsAndRoutes (893.09s)
=== CONT  TestAccIotHub_minTLSVersion
--- PASS: TestAccIotHub_identityUserAssigned (350.47s)
=== CONT  TestAccIotHub_publicAccess
--- PASS: TestAccIotHub_identitySystemAssignedUserAssigned (352.32s)
=== CONT  TestAccIotHub_customRoutes
--- PASS: TestAccIotHub_identityUpdate (667.13s)
=== CONT  TestAccIotHub_networkRulesSet
--- PASS: TestAccIotHub_minTLSVersion (299.79s)
=== CONT  TestAccIotHub_standard
--- PASS: TestAccIotHub_cloudToDevice (363.95s)
--- PASS: TestAccIotHub_endpointAuthenticationTypeUserAssignedIdentity (833.31s)
--- PASS: TestAccIotHub_LocalAuth (511.90s)
--- PASS: TestAccIotHub_endpointAuthenticationTypeUpdate (1377.94s)
--- PASS: TestAccIotHub_standard (295.57s)
--- PASS: TestAccIotHub_publicAccess (503.56s)
--- PASS: TestAccIotHub_networkRulesSet (401.75s)
--- PASS: TestAccIotHub_customRoutes (534.29s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/iothub        1
691.826s
```

an added test case
---
```
❯ tftest iothub TestAccIotHub_cosmosDBRouteUpdate
=== RUN   TestAccIotHub_cosmosDBRouteUpdate
=== PAUSE TestAccIotHub_cosmosDBRouteUpdate
=== CONT  TestAccIotHub_cosmosDBRouteUpdate
--- PASS: TestAccIotHub_cosmosDBRouteUpdate (1182.50s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/iothub        1
182.563s
```

### CODE REVIEW

**Comment by magodo** [comment]
@ziyeqf It more sounds like a user usage error as we already have the notification: https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/iothub#:~:text=Endpoints%20can%20be,is%20not%20supported:

image

**Comment by ziyeqf** [comment]
> @ziyeqf It more sounds like a user usage error as we already have the notification: https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/iothub#:~:text=Endpoints%20can%20be,is%20not%20supported:
> 
> image

no, this resource has not support cosmosdb account endpoint,  so when a user has deployed a `azurerm_iothub_endpoint_cosmosdb_account`,  then if a change has happened on `azurerm_iothub`, the update call on the iothub will remove the cosmos db endpoint. For other both supported endpoint, they are ok because there are `computed` properties for them.

**Comment by magodo** [comment]
Instead of doing this, we can split the create and update functions, and use the GET-then-PUT style in the update function, so that if users didn't specify the `endpoint`, it won't override any existing ones.

File: internal/services/iothub/iothub_resource.go [hunk 1]
```diff
@@ -793,11 +792,180 @@ func resourceIotHubCreateUpdate(d *pluginsdk.ResourceData, meta interface{}) err
...
+				Source:        utils.String(string(devices.RoutingSourceDeviceMessages)),
+				Condition:     utils.String("true"),
+				EndpointNames: &[]string{"events"},
+				IsEnabled:     utils.Bool(true),
+			}
+		}
+	}
+
+	if d.HasChange("endpoint") {
+		prop.Routing.Endpoints, err = expandIoTHubEndpoints(d, subscriptionId)
```
Comment (magodo) [inline internal/services/iothub/iothub_resource.go:879]: Still need the nil check for `prop.Routing`?

**Review by magodo — CHANGES_REQUESTED** [review]

**Review by katbyte — APPROVED** [review]
LGTM 🚡

**Pull request merged**